zokrates_core/src/static_analysis/panic_extractor.rs

```diff
@@ -0,0 +1,185 @@
+            }
+            e => fold_field_expression(self, e),
+        }
+    }
+
+    fn fold_conditional_expression<
+        E: zokrates_ast::zir::Expr<'ast, T> + Fold<'ast, T> + Conditional<'ast, T>,
+    >(
+        &mut self,
+        _: &E::Ty,
+        e: ConditionalExpression<'ast, T, E>,
+    ) -> ConditionalOrExpression<'ast, T, E> {
+        let condition = self.fold_boolean_expression(*e.condition);
+        let mut consequence_extractor = Self::default();
+        let consequence = e.consequence.fold(&mut consequence_extractor);
+        let mut alternative_extractor = Self::default();
+        let alternative = e.alternative.fold(&mut alternative_extractor);
+
+        let consequence_panics: Vec<_> = consequence_extractor.panic_buffer.drain(..).collect();
+        let alternative_panics: Vec<_> = alternative_extractor.panic_buffer.drain(..).collect();
+
+        if !(consequence_panics.is_empty() && alternative_panics.is_empty()) {
+            self.panic_buffer.push(ZirStatement::IfElse(
+                condition.clone(),
+                consequence_panics,
+                alternative_panics,
+            ));
+        }
+
+        ConditionalOrExpression::Conditional(ConditionalExpression::new(
+            condition,
+            consequence,
+            alternative,
+        ))
+    }
+
+    fn fold_uint_expression_inner(
+        &mut self,
+        b: UBitwidth,
+        e: UExpressionInner<'ast, T>,
+    ) -> UExpressionInner<'ast, T> {
+        match e {
+            UExpressionInner::Div(box n, box d) => {
+                let n = self.fold_uint_expression(n);
+                let d = self.fold_uint_expression(d);
+                self.panic_buffer.push(ZirStatement::Assertion(
+                    BooleanExpression::Not(box BooleanExpression::UintEq(
+                        box d.clone(),
+                        box UExpressionInner::Value(0).annotate(b),
+                    )),
+                    RuntimeError::DivisionByZero,
+                ));
+                UExpressionInner::Div(box n, box d)
+            }
+            e => fold_uint_expression_inner(self, b, e),
+        }
+    }
+
+    fn fold_boolean_expression(
+        &mut self,
+        e: BooleanExpression<'ast, T>,
+    ) -> BooleanExpression<'ast, T> {
+        match e {
+            e @ BooleanExpression::FieldLt(box FieldElementExpression::Number(_), _)
+            | e @ BooleanExpression::FieldLt(_, box FieldElementExpression::Number(_))
+            | e @ BooleanExpression::UintLt(
+                box UExpression {
+                    inner: UExpressionInner::Value(_),
+                    ..
+                },
+                _,
+            )
+            | e @ BooleanExpression::UintLt(
+                _,
+                box UExpression {
+                    inner: UExpressionInner::Value(_),
+                    ..
+                },
+            ) => fold_boolean_expression(self, e),
```

Comment:
Is this arm necessary? There is a default one that calls the same on line `182`

Reply:
Yes it is because in the case where in `a < b`, `a` or `b` is a constant, the comparison check is complete (works for all values) so it does not generate a panic. The next match case treats other cases where both `a` and `b` are unknown at compile time, which leads to a possible panic. If we remove this arm, then we would add an unnecessary check as these patterns would match the next case. Maybe this should be documented better in the code?

Reply:
Actually `UintLt` could be removed, but not `FieldLt`.

Reply:
Simplified and added comment